### ISSUE TITLE
invalid language string

### ISSUE THREAD

**Issue by pchampin** [w3c/rdf-turtle#37]
It just occurred to me that the following Turtle should be rejected as invalid:
```turtle
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.

[] rdfs:label "this is invalid RDF"^^rdf:langString.
```
Indeed, if accepted, it generates an invalid literal: a literal whose datatype is `rdf:langString` but which does not have a language tag. This contradicts the "if and only if" of [RDF Concepts' definition](https://www.w3.org/TR/2023/WD-rdf12-concepts-20230516/#dfn-language-tag).

Almost all implementations that I have tested accept it (notable exception: @gkellogg's Ruby implementation).

We probably need a negative test for this, and of course similar tests for other concrete syntaxes.

**Comment by afs** [comment]
Special case datatypes are problematic. e.g. `STRDT` e.g. subclasses/derived types of that datatype.

It can be considered an ill-formed literal according to the datatype, much like `"tuesday"^^xsd:integer`. 

The text below the defn: "A literal is a language-tagged string if the third element is present. "  -  says it isn't a language-tagged string i.e. it has no value (the "literal value" text).

(Also -- bullet two only refers to lexical form.)

**Comment by gkellogg** [comment]
> Almost all implementations that I have tested accept it (notable exception: @gkellogg's Ruby implementation).

My implementation does, in fact, parse it, even though it's invalid, as it parses `"tuesday"^^xsd:integer`. It's when serializing to N-Triples that an error is generated (presuming you're using my [distiller](http://rdf.greggkellogg.net/distiller?command=serialize&format=turtle&output_format=turtle)). If you change the `output_format` to `turtle`, you should get the input back out. There's also a `validate` option that would cause the parse phase to fail with errors.

My interpretation is that it's legal Turtle, as @afs says, it just results in an invalid triple. That would be the case for language tags which are also not valid according to BCP47.

**Comment by pchampin** [comment]
> It can be considered an ill-formed literal according to the datatype, much like `"tuesday"^^xsd:integer`

This is very different, IMO. `"tuesday"^^xsd:integer` is "a semantic inconsistency but is not *syntactically* ill-formed" (quoting rdf-concepts). On the other hand, a literal with datatype `rdf:langString` and no (or an empty) language tag is syntactically wrong.

On a practical level, it means that many implementation will choke or behave strangely with it. As @gkellogg points out, his implementation will crash when trying to serialize it back to N-Triples. It will not crash when serializing it back to Turtle, but will produce invalid Turtle, namely: `"this is invalid RDF"@` (empty language tag).

**Comment by afs** [comment]
> means that many implementation will choke or behave strangely with it.

Your testing showed this isn't the case. Nothing has changed from RDF 1.1.

> should be rejected as invalid:

At one level, if it's wrong, then it's outside the spec and we don't define the behaviour. We might suggest a behaviour but that's not the same as requiring a behaviour.

There are good reasons for systems to choose to accept it - if steaming large data, one occurrence far down the input stream, throwing an error and aborting a large load after a few hours is extremely inconvenient.

**Comment by pchampin** [comment]
> > means that many implementation will choke or behave strangely with it.
> 
> Your testing showed this isn't the case.

My testing shows that most implementation don't choke on it., that's right. I do expect that they behave strangely down the line, though. E.g., @gkellogg 's implementation accepts it during parsing, but then fails to serialize it back...

> Nothing has changed from RDF 1.1.

Indeed. I'm not claiming that this issue is specific to RDF 1.2.

> > should be rejected as invalid:
> 
> At one level, if it's wrong, then it's outside the spec and we don't define the behaviour. We might suggest a behaviour but that's not the same as requiring a behaviour.

I agree that we do not define the behaviour of parsers when they encounter invalid input.
Let me rephrase my suggestion then: we should change the Turtle syntax to make my example above invalid Turtle. As a rule, I think we must not allow for something that is valid in a concrete syntax, but has no counterpart in the abstract syntax. This is what happens here.

An alternative would be to relax the constraints on `rdf:langString`  in the abstract syntax, so that an a literal of this datatype could exist (in the abstract syntax) without a language tag (although they would be considered ill-formed, and be therefore inconsistent in the semantics). This would make the example above syntactically (concrete *and* abstract) valid.

> There are good reasons for systems to choose to accept it - if steaming large data, one occurrence far down the input stream, throwing an error and aborting a large load after a few hours is extremely inconvenient.

Agreed, but this is a very general consideration, not specific to this issue.

**Comment by afs** [comment]
Of the two options, I prefer the option to relax the abstract syntax and have it be ill-formed. i.e. widen and not restrict existing usage.

[JSON-LD already says](https://www.w3.org/TR/json-ld11/#terminology) "and, if the datatype is rdf:langString, an optional language tag" and systems do accept it.

I'd prefer leaving things as they are.

JSON-LD and RDF/XML don't rely on the document parsing for language tags.

(The title of the issue maybe should be "missing language string" for the long term record of the WG's work.)

**Comment by afs** [comment]
Here's another corner case for "missing language tag".

```turtle
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

[] rdfs:label "abc"@--rtl .
```

Sort of "unknown language but known to be rtl" or "always rtl".

**Comment by afs** [comment]
Seen in the wild: https://github.com/apache/jena/issues/2555 

That's in a result set but the principle is the same.

**Comment by afs** [comment]
> all primary language subtags are two or three characters.

... currently ... RFC 5646 allows more:

```ebnf
 language      = 2*3ALPHA            ; shortest ISO 639 code
                 ["-" extlang]       ; sometimes followed by
                                     ; extended language subtags
               / 4ALPHA              ; or reserved for future use
               / 5*8ALPHA            ; or registered language subtag
```

**Comment by gkellogg** [comment]
> I think that before closing this, we need to decide whether `"foo"^^xsd:rdf:langString` is
> 
> * invalid in the abstract syntax or
> * valid in the abstract syntax but ill-formed.
> 
> If we chose option 1, we don't need IMO to change the text in RDF-Concepts, but we need negative tests in Turtle, N-Triples, etc. If we chose option 2, we need to update the text in RDF-Concepts. We probably also need additional tests in RDF-Semantics, because we now have a new way to express RDF-inconsistent graphs (which, incidentally, was not possible before, AFAIK...).

Yes, Concepts is explicit about requirements for using `rdf:langString` and `rdf:dirLangString`, so these could be added to the processing rules. The handler for [RDFLiteral](handle-RDFLiteral) would need to  exclude either `rdf:langString` and `rdf:dirLangString` from forming a datatype IRI creating a processor error. This would need to be done the same for the N-Triples [literal](handle-literal) handler. Both handlers probably also need to enforce the BCP47 [well-formedness](https://www.rfc-editor.org/rfc/rfc5646#section-2.2.9) conditions and reject non-conforming language tags.

We also allow arbitrary alphabetic sequences to be used as base direction, although Concepts limits it to "ltr" and "rtl".

My parsers accept these forms without validation. Validation can be done at the same time, or by looking at the graph for inconsistencies such as language tags and base direction, which is more of a semantic constraint, rather than syntactic.

The question becomes, where to stop. In other cases, we look for ill-formed literals in semantics, rather than when parsing (e.g., `"MMXXV"^xsd:date`), so this is a slippery slope.

**Comment by afs** [comment]
I don't think having negative tests is particularly helpful in cases where common mistake happen but the Turtle syntax is right - it is implying "stop parsing" and when streaming. Bad language tags happen.

Apache Jena:

`:x1 :p "foo"^^rdf:langString .` ==> Warning. I have seen this in the wild. In large data, junk happens. Stopping parsing at that point (error) is unhelpful when checking.

`:x2 :p "foo"^^rdf:dirLangString` ==> Warning.

`:x3 :p "foo"@--ltr` => Error (it is only justifiable because this is a new feature).

`:x4 :p "foo"@a` ==> Warning (invalid language tags are not unusual)

The spec text would be clearer if it used MUST on presence of these components (which I can PR but not right this moment):

"a non-empty language tag as defined by [[BCP47](https://www.w3.org/TR/rdf12-concepts/#bib-bcp47)] MUST be present"

"a base direction MUST be present and MUST be one of the following:"

**Comment by gkellogg** [comment]
I'm happy to stick with warnings for these things. Probably semantic validation is the place to catch it, or with an appropriate implementation-specific parser flag.

Regarding base direction, we intentionally used `'--' [a-zA-Z]+` to keep this open with the idea of checking elsewhere. Otherwise, LANG_DIR could be updated to tie this down: `'--' ('ltr' | 'rtl')`. There is no data in the wild for this quite yet, so the abuse of the language tag requiring a more lenient parser rule makes more sense there.

**Comment by pchampin** [comment]
@gkellogg 
> by looking at the graph for inconsistencies such as language tags and base direction, which is more of a semantic constraint, rather than syntactic.

I disagree: this is a violation of the abstract syntax as defined by RDF-Concepts, not of a semantic condition defined by RDF-Semantics...

@afs
> I don't think having negative tests is particularly helpful in cases where common mistake happen but the Turtle syntax is right - it is implying "stop parsing" and when streaming. Bad language tags happen.

I think we are mixing two different issues here:
* should we forbid parsers to produce data that does not comply with the abstract syntax, even when the input text complies with the concrete syntax grammar? 
* should we encourage parsers to continue parsing when valid input would lead to invalid (abstract) syntax?

I do think that we should answer "yes" to both questions...

**Comment by afs** [comment]
@pchampin 

There is "complies" and "does not comply".

"Forbid" is not something in RDF 1.1 and adding negative syntax tests would mean RDF 1.1 systems would stop passing the test suite on RDF 1.1 data. The only way I can see for your two points to be "yes" is to expect parsers switch to "no output after constraint violation" mode. That is a new requirement - maybe some do, but not all do as of today. It would forbid error recovery for example.

Sometimes it's someone else's data.  Wikidata can have bad language tags (9 character subtags). Failing after a few hours loading on a point technicality ("forbid") is not a good experience. It does not comply with the specs but telling system what to do about it is another step.

It is better IMO to have a test suite that implementers target and users run, not with some `--spec` flag but with the parser in the normal deployment mode.

**Comment by gkellogg** [comment]
Generally, reporting on such an error, continuing to parse the document and not emitting any triples is problematic at least for streaming serializers. This might require an implementation to do some form of commit/fallback logic to stage the changes into a buffer to either commit or rollback if errors are found. But, the situation is not unlike any error which is found during processing, and it's usually sufficient to just emit triples that aren't in error. So in this case, a literal with a missing or ill-formed language tag would not cause that specific triples to be emitted, but others before and after would.

We often tests these things with negative syntax tests, but it might require a positive evaluation test that looks for all triples, except those with an error. Still, this would depend on consistent behavior between parsers, which likely doesn't really exist.

**Comment by afs** [comment]
> and it's usually sufficient to just emit triples that aren't in error.

I disagree. Bad language tags is a concrete example where loading then correcting is

Skipping bad triples is in itself risky especially at scale. Intended information has been lost.

**Comment by TallTed** [comment]
I feel strongly that mis-typed literals — including those with bad langtags or no langtag while typed as langString — should be loaded from serialization as if they were appropriately escaped to make the *whole string* an untyped literal.

I also feel that "bad data", including nonconformant triples of nearly if not all variation, should be loaded with an alert of some kind (app log, system log, modal or not user interaction, etc.), possibly putting such nonconformant data into a special table or graph within the data store, such that the user can make informed judgements about their handling. 

I would be happy to have a loading flag that says "drop bad data" (or specified other behavior when encountering bad data, such as a sidecar table, DB, or serialization file), as long as that is not the default (unset) behavior, and providing that wherever such "bad data" gets written, it can be reasonably human-reviewed (so, including any necessary PREFIX, BASE, etc., declarations, and that, when loading Turtle or other sugar-filled serializations, enough of the "surrounding" sugar is preserved for such human-cleaning to be possible — e.g., Turtle that uses the `,` and `;` features would have to preserver the entire stanza from `<SUBJECT>` through to the final `.`).

(These feelings are not necessarily reflected today in Virtuoso or other tools with which I've been associated, but if they were reflected in the RDF & SPARQL standards, their implementation would be far more likely.)
